File: sample/src/main/java/com/matchday/ui/MatchUiState.kt
package com.matchday.ui

import kotlinx.serialization.Serializable
import kotlinx.serialization.Transient

@Serializable
data class MatchUiState(
    val home: Team,
    val away: Team,
    val homeGoals: Int = 0,
    val awayGoals: Int = 0,
    private val started: Boolean = false,
    private val running: Boolean = false,
    private val finished: Boolean = false
) {

    @Transient
    val ongoing: Boolean = started && running && !finished

    @Transient
    val paused: Boolean = started && !running && !finished

    @Transient
    val ended: Boolean = started && finished

    fun startGame(): MatchUiState = copy(started = true, running = true, finished = false)
    fun pausedGame(): MatchUiState = copy(started = true, running = false, finished = false)
    fun endGame(): MatchUiState = copy(started = false, running = false, finished = true)

    fun lineup() = "${home.country} ${home.flag} vs ${away.flag} ${away.country}"
    fun score() = "$homeGoals - $awayGoals"
}

@Serializable
data class Team(
    val country: String,
    val flag: String
)

File: sample-ios/MatchDay/Screens/MatchUiState.swift
import Foundation

struct Team: Codable {
    let country: String
    let flag: String
}

struct MatchUiState: Codable {
    let home: Team
    let away: Team
    var homeGoals: Int = 0
    var awayGoals: Int = 0
    private var started: Bool = false
    private var running: Bool = false
    private var finished: Bool = false
    
    var ongoing: Bool { started && running && !finished }
    var paused: Bool { started && !running && !finished }
    var ended: Bool { started && finished }
    
    init(home: Team, away: Team) {
        self.home = home
        self.away = away
    }
    
    private mutating func updateGameState(started: Bool, running: Bool, finished: Bool) -> MatchUiState {
        var newState = self
        newState.started = started
        newState.running = running
        newState.finished = finished
        return newState
    }
    
    mutating func startGame() -> MatchUiState {
        return updateGameState(started: true, running: true, finished: false)
    }
    
    mutating func pausedGame() -> MatchUiState {
        return updateGameState(started: true, running: false, finished: false)
    }
    
    mutating func endGame() -> MatchUiState {
        return updateGameState(started: false, running: false, finished: true)
    }
    
    func lineup() -> String {
        return "\(home.country) \(home.flag) vs \(away.flag) \(away.country)"
    }
    
    func score() -> String {
        return "\(homeGoals) - \(awayGoals)"
    }
}
